### ISSUE TITLE
setExternalUserId() with Identity Verification : invalid identifier_auth_hash

### ISSUE THREAD

**Issue by CaM2091** [OneSignal/react-native-onesignal#1122]
**Description:**

I'm trying to set external user id with setExternalUserId() using Identity Verification.
My backend generate an auth hash like this : 

```
const generateAuthHash = (email: string) => {
  const hmac = crypto.createHmac("sha256", process.env.ONESIGNAL_REST_API_KEY);
  hmac.update(email);
  return hmac.digest("hex");
};

const authHash = generateAuthHash("test@example.com")
```

When I use the generated auth hash to setEmail() on my app, everything is ok. The auth hash seems valid.

But when I used auth hash to setExternalUserId(), I have an error : 

```
OneSignal.setLogLevel(6, 6);
OneSignal.setExternalUserId(
  "MyDBUserId",
  "MyAuthHashFromBackend"
  (results: any) => {
    console.log('Results of setting external user id');
    console.log(results);
    if (results.push && results.push.success) {
      console.log(
        'Results of setting external user id push status:',
      );
      console.log(results.push.success);
    }
    if (results.email && results.email.success) {
      console.log(
        'Results of setting external user id email status:',
      );
      console.log(results.email.success);
    }
  },
)
```

![image](https://user-images.githubusercontent.com/6414269/102373256-4cd17c00-3fc0-11eb-8763-d2f9ea07df38.png)

![IMG!UNITO-UNDERSCORE!0464](https://user-images.githubusercontent.com/6414269/102372071-219a5d00-3fbf-11eb-9091-0955fecd591b.jpg)

I don't understand, is auth hash different for setEmail() and setExternalUserId() ?

**Environment**

"react-native": "0.62.1"
"react-native-onesignal": "^3.9.3"

**Steps to Reproduce Issue:**

1. Setup OneSignal SDK for React Native
1. Try set external user id with Identity Verification

**Comment by CaM2091** [comment]
Ok I found solution.
Setting external user id require specific auth hash generated with the DB user id, like this :

```
const generateAuthHash = (data: string) => {
  const hmac = crypto.createHmac("sha256", process.env.ONESIGNAL_REST_API_KEY);
  hmac.update(data);
  return hmac.digest("hex");
};

const emailAuthHash = generateAuthHash("test@example.com")
const externalIdAuthHash = generateAuthHash("MyDBUserId")
```

Documentation only show example for email.

But now I have another issue : the external user Id is not setted on both channels (Email & Push), why ?
Moreover, when I update tags from app side, tags are setted on both channels. But when I update tags from backend with editTagsWithExternalUserIdDevice(), tags are only setted in one channel. Is there any reason ?

![Capture_d’écran_2020-12-16_à_17 58 31](https://user-images.githubusercontent.com/6414269/102381810-f79a6800-3fc9-11eb-9ad0-6aed875f5532.jpg)

Thanks for your help

**Comment by rgomezp** [comment]
Howdy,

[Reference](https://documentation.onesignal.com/docs/identity-verification)
---

I have updated the docs to include an example for the external id. It's really the same thing just using `identifier` instead of `email`.

Regarding the external id being set in both channels, I got the following from setting via the app only:

![Screen Shot 2020-12-16 at 1 26 27 PM](https://user-images.githubusercontent.com/11739227/102396647-5946dc00-3fa2-11eb-8edb-e5752fc9606e.png)

As you can see from the above screenshot, both the external id and tags are set on both. The only value that shouldn't show on both is `email` interestingly. I realize that might be confusing. We will try to fix this discrepancy in the future. We plan to unify channels under a single model such as "Person" or "User." But that's neither here nor there.

Your code
```
const generateAuthHash = (data: string) => {
  const hmac = crypto.createHmac("sha256", process.env.ONESIGNAL_REST_API_KEY);
  hmac.update(email);
  return hmac.digest("hex");
};
```
the function `generateAuthHash` takes a variable `data` that actually never gets used inside the function. Please double check that you're using the correct variable there. I'm surprised Typescript didn't actually warn you about this since it looks you're trying to use the variable `email`. Assuming this is just a reporting mistake on your part, I'll assume you managed to successfully set the correct auth hash for the following...

From the backend, are you setting tags and external id using the **push player** id and not the **email record player** id?

We will also try reproducing by setting some tags via the API. In the meantime, if you could please provide some more clarity regarding which player record you're updating via the REST API (e.g: push player id, email id, or external id). We would love to see some code. Make sure to obfuscate your secrets.

Cheers

**Comment by CaM2091** [comment]
Hello @rgomezp !
Thanks for your answer, I will provide more informations later :) 
(And yes, for `generateAuthHash`, it was just a reporting mistake)

For the moment, I'm doing some test, and I noticed my 2 players ids are different : 

<img width="575" alt="Capture d’écran 2020-12-17 à 11 37 35" src="https://user-images.githubusercontent.com/6414269/102477394-98d1fe80-405c-11eb-8b70-53c337f35c51.png">

Is it normal ? I think it can be linked to my issue, but not sure

<img width="1188" alt="Capture d’écran 2020-12-17 à 12 26 18" src="https://user-images.githubusercontent.com/6414269/102482293-45af7a00-4063-11eb-9ad6-1f6d5d5b07db.png">
Sometimes, there is no device name 🤔

**Comment by CaM2091** [comment]
So, for my debug test, I take your code on https://github.com/OneSignal/react-native-onesignal/blob/master/examples/RNOneSignal/App.js and I add a call to my backend server to get externalId and email auth hash. I tried on iPhone X, iPhone 6 and Huawei P20.  Always the same problem, external user id is only set in one channel 😩

<img width="424" alt="Capture d’écran 2020-12-17 à 15 54 31" src="https://user-images.githubusercontent.com/6414269/102504039-a4372100-4080-11eb-8dbc-446fc0ca32d8.jpeg">

<img width="424" alt="Capture d’écran 2020-12-17 à 15 54 31" src="https://user-images.githubusercontent.com/6414269/102503621-32f76e00-4080-11eb-95eb-616faf1b5333.png">

<img width="1353" alt="Capture d’écran 2020-12-17 à 16 01 31" src="https://user-images.githubusercontent.com/6414269/102504681-6ab2e580-4081-11eb-9c79-8bd53e8158af.png">

Edit : This package https://github.com/zo0r/react-native-push-notification is used in my app, It can be linked, I try remove it and tell you

Edit 2 : I removed react-native-push-notification and reinstall everything, it changes nothing :/

Edit 3 : I restart from a fresh project using example project (https://github.com/OneSignal/react-native-onesignal) (just adding call to my backend auth hash generation method for Identity Verification) and exactly same problem... I don't understand...

**Comment by rgomezp** [comment]
Howdy,
From your log, I can tell that the call itself is not succeeding ("success": 0).

That is why it is is not working for the email player.

Can you see what is actually being printed by the native logs when that happens? The logs in Xcode should have the full response from our backend systems.

**Comment by rgomezp** [comment]
Closing due to no response

**Comment by uonuon** [comment]
https://documentation.onesignal.com/docs/identity-verification
Make sure you follow docs, 
```
let externalUserId = '123456789'; // You will supply the external user id to the OneSignal SDK
let externalUserIdAuthHash = "..."; // Identifier auth hash generated from your server

// Setting External User Id with Callback Available in SDK Version 3.9.3+
OneSignal.setExternalUserId(externalUserId, externalUserIdAuthHash, (results) => {
  // The results will contain push and email success statuses
  console.log('Results of setting external user id');
  console.log(results);
  
  // Push can be expected in almost every situation with a success status, but
  // as a pre-caution its good to verify it exists
  if (results.push && results.push.success) {
    console.log('Results of setting external user id push status:');
    console.log(results.push.success);
  }
  
  // Verify the email is set or check that the results have an email success status
  if (results.email && results.email.success) {
    console.log('Results of setting external user id email status:');
    console.log(results.email.success);
  }
});
```
